### PR TITLE
Shengwei create setup history modal

### PR DESCRIPTION
**Pull request by ptrpengdev**
# Description
image

## Related PRS (if any):
This backend PR is related to the frontend [PR1992.](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1992)

## Main changes explained:
- Add fetch and update (refresh, cancel) setup token API in profileInitialSetupController.js
- Update weekly task: deleteExpiredTokens to allow 90-day data retention for expired or cancelled tokens.

## How to test:
1. check into current branch and the frontend [PR1992.](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1992)
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. Setup backend Gmail Oauth and .env in the backend: [Link for video tutorial ](https://docs.google.com/document/d/1HJJTDtWykjGb4IfiIFlHXDfp_IABvmwKxsMPqqF6rG8/edit) and update local env file
5. Refer to frontend [PR1992.](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1992) for testing

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/69882989/6c86eb1e-8d0b-42db-9c15-28b18c81366b

## Note:
You need to Setup gmail oauth feature for testing. Also, change the email variable in the **refreshSetupInvitation** function in **src/controllers/profileInitialSetupController.js** (backend). Comment out `const { email } = result;` and use `const email = 'YOUR_EMAIL_TO_RECEIVE_MSG@gmail.com';` so that the application will send email to the email you defined!

### CODE REVIEW

**Review by skaush21 — COMMENTED** [review]
* logged in using owner account
* went to 'other links' -> 'user management' -> 'bell icon'
* tested each of the drop down menus and the search query for emails, works
* clicked cancel, checked email, did not receive an email
* clicked refresh, checked email, received an email, checked link, link works
* repeated previous steps, clicking cancel does not send an email.
https://github.com/OneCommunityGlobal/HGNRest/assets/141872402/70abccf7-cfef-4d6c-8dcd-210f0a383cb5
I took a look at the code in _profileInitialSetupController.js_ I noticed that _refreshSetupInvitation_ has a segment of code that sends an email when it is being refreshed. this feature only affects when it is being refreshed. Above this segment of code there's a _cancelSetupInvitation_ which does not include the code needed to send an email: 
`
const link = `${baseUrl}/ProfileInitialSetup/${result.token}`;
sendEmailWithAcknowledgment(
             email,
            'NEEDED: Complete your One Community profile setup',
            sendLinkMessage(link),
          );
`
As shown in the video, the cancel button is suppose to send an email, but in your code, it's setup to not show the email. Is this intentional?
Screen Shot 2024-02-29 at 10 50 31 AM

**Review by skaush21 — APPROVED** [review]
features implemented work as per the description.

**Comment by 20chen-7** [comment]
left review at the frontend, but will continue review later for double check invitation history

**Review by abdel-lall — COMMENTED** [review]
Left a comment on the frontend.

File: src/controllers/profileInitialSetupController.js [hunk 1]
```diff
@@ -131,11 +143,14 @@ const profileInitialSetupController = function (
...
           expiration: expiration.toDate(),
+          isSetupCompleted: false,
+          isCancelled: false,
+          createdDate: Date.now(),
         });
 
         const savedToken = await newToken.save();
         const link = `${baseUrl}/ProfileInitialSetup/${savedToken.token}`;
-
+        console.log(baseUrl, savedToken.token);
```
Comment (abdel-lall) [inline src/controllers/profileInitialSetupController.js:153]: please remove the test code.

File: src/controllers/profileInitialSetupController.js [hunk 2]
```diff
@@ -163,13 +179,19 @@ const profileInitialSetupController = function (
       if (foundToken) {
         const expirationMoment = moment(foundToken.expiration);
 
-        if (expirationMoment.isAfter(currentMoment)) {
+        if (foundToken.isSetupCompleted) {
+          res.status(400).send(TOKEN_HAS_SETUP_MESSAGE);
+        } else if (foundToken.isCancelled) {
+          res.status(400).send(TOKEN_CANCEL_MESSAGE);
+        } else if (expirationMoment.isAfter(currentMoment)) {
```
Comment (abdel-lall) [inline src/controllers/profileInitialSetupController.js:186]: It's advisable to place the success condition as the final one.

File: src/controllers/profileInitialSetupController.js [hunk 3]
```diff
@@ -186,7 +208,7 @@ const profileInitialSetupController = function (
  - Sends the JWT as a response.
 */
   const setUpNewUser = async (req, res) => {
-    const { token } = req.body;
+    let { token } = req.body;
```
Comment (abdel-lall) [inline src/controllers/profileInitialSetupController.js:211]: This is posing issues for the initial setup functionality. The setup token differs from the JWT authentication token. Please consider renaming one of them. And, ensure that the JWT sent to the frontend is an object with a property named 'token'.

File: src/controllers/profileInitialSetupController.js [hunk 4]
```diff
@@ -275,7 +300,7 @@ const profileInitialSetupController = function (
           };
 
           res.send({ token }).status(200);
-
+          ProfileInitialSetupToken.update();
```
Comment (abdel-lall) [inline src/controllers/profileInitialSetupController.js:303]: No arguments were passed to the update function. Was it intended to update the 'isSetupCompleted' property?

**Review by abdel-lall — CHANGES_REQUESTED** [review]

File: src/controllers/profileInitialSetupController.js (hunk 3 above)
Comment (ptrpengdev) [reply src/controllers/profileInitialSetupController.js:211]: Modifying the data payload structure for both request and response falls outside the scope of this pull request. The token is used for carrying new user setup token and JWT across different functions/APIs. Should document this as an improvement in the HGN document.

File: src/controllers/profileInitialSetupController.js (hunk 4 above)
Comment (ptrpengdev) [reply src/controllers/profileInitialSetupController.js:303]: Correct. Will include the updated code in next commit.

File: src/controllers/profileInitialSetupController.js (hunk 2 above)
Comment (ptrpengdev) [reply src/controllers/profileInitialSetupController.js]: Updated.

File: src/controllers/profileInitialSetupController.js (hunk 1 above)
Comment (ptrpengdev) [reply src/controllers/profileInitialSetupController.js]: Removed.

File: src/controllers/profileInitialSetupController.js (hunk 3 above)
Comment (abdel-lall) [reply src/controllers/profileInitialSetupController.js]: Apologies for any confusion. I wanted to highlight a concern regarding the modifications made: changing from const { token } = req.body; to let { token } = req.body;, and subsequently using the same variable for the JWT on line 291. It appears that these two tokens serve different purposes. Therefore, I suggest considering renaming one of them.

Comment (ptrpengdev) [reply src/controllers/profileInitialSetupController.js]: Updated.        
const { token } = req.body;
...
const jwtToken = jwt.sign(jwtPayload, JWT_SECRET);
 ...
res.send({ token: jwtToken }).status(200);

File: src/controllers/profileInitialSetupController.js [hunk 5]
```diff
@@ -156,20 +170,25 @@ const profileInitialSetupController = function (
...
+        if (foundToken.isSetupCompleted) {
+          res.status(400).send(TOKEN_HAS_SETUP_MESSAGE);
+        } else if (foundToken.isCancelled) {
+          res.status(400).send(TOKEN_CANCEL_MESSAGE);
+        } else if (expirationMoment.isBefore(currentMoment)) {
+          res.status(400).send(TOKEN_EXPIRED_MESSAGE);
         } else {
-          res.status(400).send('Invalid token');
-        }
+          res.status(400).send(TOKEN_INVALID_MESSAGE);
```
Comment (abdel-lall) [inline src/controllers/profileInitialSetupController.js:187]: Hi @ptrpengdev, I wanted to bring to your attention that the token will always be invalid. You've set the default condition to res.status(400).send(TOKEN_INVALID_MESSAGE)

Comment (ptrpengdev) [reply src/controllers/profileInitialSetupController.js]: My mistatke. Updated.

**Review by cgomezhub — CHANGES_REQUESTED** [review]
I left comment on  frontend PR  #1992

https://www.loom.com/share/3b4c464352eb4356b12864e5c74b50d7?sid=609bc7aa-05cf-4fdf-9a10-7b61024ebba5

**Review by cgomezhub — APPROVED** [review]
I left a new comment on[ PR 1992](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1992#pullrequestreview-2048609913)

https://www.loom.com/share/d15b825e33644d0aa9ed98adfa6d814b?sid=3fcc744a-594b-4d08-943f-619dc8222156

**Review by abdel-lall — APPROVED** [review]

**Comment by one-community** [comment]
Thank you all, merging!

**Pull request merged**